adaptive_memory_v3.1.py

```
@@ -112,6 +114,10 @@ class Filter:
     _memory_embeddings = {}
     _relevance_cache = {}
 
+    # Class-level locks for message deduplication (shared across all instances)
+    _message_locks = {}  # Dict of message_hash -> asyncio.Lock
```

Ruff (0.14.10)

Mutable class attributes should be annotated with `typing.ClassVar`

(RUF012)